### PR TITLE
soundwire: bus: initialize bus clock base and scale registers

### PR DESCRIPTION
**Pull request by plbossart**
The SoundWire 1.2 specification adds new registers to allow for
seamless clock changes while audio transfers are on-going. Program
them following the specification.

Note that dynamic clock changes are not supported for now, this only
adds the register initialization.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

### CODE REVIEW

**Comment by plbossart** [comment]
@shumingfan @oder-chiou FYI. Those base/scale registers should not be part of the regmap.

File: drivers/soundwire/bus.c [hunk 1]
```diff
@@ -1082,12 +1082,109 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
...
+
+	scale = mclk_freq / curr_freq - base_offset;
+
+	/* map scale to Table 90 of SoundWire 1.2 spec */
+	for (i = 0; i < 7; i++) {
+		if (scale == BIT(i)) {
+			scale_index = i + 1;
+			break;
+		}
+	}
```
Comment (ranj063) [inline drivers/soundwire/bus.c]: @plbossart can we use ffs() here?

Comment (plbossart) [reply drivers/soundwire/bus.c]: @ranj063 I thought about it but I really want to check for powers of 2 only. All other cases are errors. With ffs() you would get the first non-zero bit, not that there is a single bit set.

Comment (ranj063) [reply drivers/soundwire/bus.c]: I see. maybe is_power_of_2() then?

Comment (plbossart) [reply drivers/soundwire/bus.c]: Maybe. This is going to be multiple tests anyways since I also need to check the value is 64 or lower. The loop is more brute force but there's a single check.
I am sure @lyakh will have an optimization from hell on this one :-)

File: drivers/soundwire/bus.c [hunk 2]
```diff
@@ -1082,12 +1082,109 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
...
+		return 0;
+
+	if (!mclk_freq) {
+		dev_err(&slave->dev,
+			"no bus MCLK, cannot set SDW_SCP_BUS_CLOCK_BASE\n");
+		return -EINVAL;
+	}
+
+	/* map base frequency using Table 89 of SoundWire 1.2 spec */
+	if (!(38400000 % mclk_freq)) {
```
Comment (RanderWang) [inline drivers/soundwire/bus.c]: if mclk_freq = 600000 or 8000000,  only last if will take effect ?

File: drivers/soundwire/bus.c [hunk 3]
```diff
@@ -1082,12 +1082,109 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
...
+	/* paranoia checks */
+	if (!base) {
+		dev_err(&slave->dev,
+			"Unsupported clock base\n");
+		return -EINVAL;
+	}
+
+	if (mclk_freq % curr_freq) {
+		dev_err(&slave->dev,
+			"MCLK and bus curr_freq are not multiple\n");
```
Comment (RanderWang) [inline drivers/soundwire/bus.c]: Can we print frequency value to make it more instructive ?

File: drivers/soundwire/bus.c (hunk 1 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c]: I am eager for your reply. @lyakh

Comment (bardliao) [reply drivers/soundwire/bus.c]: Maybe
```
if (scale == BIT(bits_per(scale) - 1) && scale <= 64)
    scale_index = bits_per(scale);
else
    dev_err()
```
But it is less readable, so how about
```
switch (scale) {
    case 1:
    case 2:
    case 4:
    case 8:
    case 16:
    case 32:
    case 64:
        scale_index = bits_per(scale);
        break;
    default:
        dev_err()
        return -EINVAL;
```

File: drivers/soundwire/bus.c [hunk 4]
```diff
@@ -1082,12 +1082,109 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
...
+	if (!mclk_freq) {
+		dev_err(&slave->dev,
+			"no bus MCLK, cannot set SDW_SCP_BUS_CLOCK_BASE\n");
+		return -EINVAL;
+	}
+
+	/* map base frequency using Table 89 of SoundWire 1.2 spec */
+	if (!(38400000 % mclk_freq)) {
+		base = SDW_SCP_BASE_CLOCK_19200000_HZ;
+		base_offset = 1;
```
Comment (kv2019i) [inline drivers/soundwire/bus.c]: Hmm, this is slightly hard to follow. If we assume sane mclk_freq value range, then the algorithm seems to work though, so ok. E.g. mclk_freq of 0.6Mhz would lead to base_offset=1 and base=24Mhz, but that would fail in trying to find a suitable scaling factor. So looks ok, although a bit hard to review.

File: drivers/soundwire/bus.c (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/bus.c]: mclk_freq is only going to be 38.4 MHz or 24 MHz for intel. the rest is just in case someone needs other values.

File: drivers/soundwire/bus.c (hunk 4 above)
Comment (plbossart) [reply drivers/soundwire/bus.c]: mclk is a property in the BIOS, it's currently:

"intel-sdw-ip-clock", 
0x016E3600 -> 24
0x0249F000 -> 38,4

The base_offset 1 is because MIPI defines 19.2 so we have a factor to correct.
Will add some more comments here.

File: drivers/soundwire/bus.c (hunk 1 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c]:
````
factor = ffs(scale);
if (!is_power_of_2 (scale) || factor >= 8)
       dev_err;
````

File: drivers/soundwire/bus.c [hunk 5]
```diff
@@ -1082,12 +1082,109 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
...
+	}
+	if (!(24000000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_24000000_HZ;
+	if (!(24576000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_24576000_HZ;
+	if (!(22579200 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_22579200_HZ;
+	if (!(32000000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_32000000_HZ;
```
Comment (lyakh) [inline drivers/soundwire/bus.c:1131]: just to make sure: every following one is better than all previous ones, right? Because several of them might trigger. But in fact, say, if mclk_freq is a factor of both 32MHz and 24MHz, a base of 32MHz will be selected. Why is it better than a base of 24MHz? At least 24MHz should consume less?

File: drivers/soundwire/bus.c (hunk 3 above)
Comment (lyakh) [reply drivers/soundwire/bus.c]: not being a native English speaker - the above sentence means the same as "MCLK is not a multiple of curr_freq?"

File: drivers/soundwire/bus.c (hunk 1 above)
Comment (lyakh) [reply drivers/soundwire/bus.c]: @plbossart @RanderWang I feel flattered, but my otherwise direct super-low-latency link to the hell seems broken today :-) But no, I don't think I can much improve what already has been proposed. Something like this I guess:
```
scale_idx = ilog2(scale) + 1;
if (BIT(scale_idx - 1) != scale || scale_idx > 7) {
	/* error */
}
```
Using `ffs()` above instead of `ilog2() + 1` could be very slightly more optimal because usually the value would be within the low 7 bits but using `ilog2()` seems more explicit here, which is something we usually prefer ;-) Otherwise either of them should be fine.
Also, using one of these versions would eliminate the need to initialise `scale_idx`

File: drivers/soundwire/bus.c (hunk 5 above)
Comment (plbossart) [reply drivers/soundwire/bus.c:1131]: There is ONE mclk_freq provided by firmware. No one is trying to optimize to find the best clock, just trying to use firmware-provided information and convert into SoundWire 1.2 register values.

File: drivers/soundwire/bus.c (hunk 3 above)
Comment (plbossart) [reply drivers/soundwire/bus.c]: fixed

File: drivers/soundwire/bus.c (hunk 5 above)
Comment (bardliao) [reply drivers/soundwire/bus.c:1131]: In other words, this is to inform Slave the clock frequency not to set the clock frequency.

Comment (lyakh) [reply drivers/soundwire/bus.c:1131]: Above code calculates the `base` variable, which is then written to the `SDW_SCP_BUS_CLOCK_BASE` register. Let's say `mclk_freq == 8000000`. The code will first enter
```
if (!(24000000 % mclk_freq))
	base = SDW_SCP_BASE_CLOCK_24000000_HZ;
```
and thus set `base` to 24MHz. But then it'll also enter
```
if (!(32000000 % mclk_freq))
	base = SDW_SCP_BASE_CLOCK_32000000_HZ;
```
and rewrite `base` to 32MHz. Is this intended? As far as I understand both 24MHz and 32MHz would work, right? Why do we then choose 32MHz? I don't know what that register does, apart from what the print below says, i.e. that it "configures the bus base," but if it does set a clock frequency to the respective value, then the lower frequency might be preferred to save power.

Comment (plbossart) [reply drivers/soundwire/bus.c:1131]: no, as described in the code we don't select a clock, just mirror the settings at the Slave level.

**Comment by plbossart** [comment]
updated version with all previous feedback addressed, but let's not merge it just yet, I still need to verify it on hardware.

**Comment by plbossart** [comment]
Removing RFC since I tested this with a SoundWire 1.2 device with CNL and found the right values:
````
[   26.627355] rt711-sdca sdw:0:25d:711:1: Configured bus base 2, scale 3, mclk 24000000, curr_freq 6000000
````

File: drivers/soundwire/bus.c (hunk 5 above)
Comment (lyakh) [reply drivers/soundwire/bus.c:1131]: @plbossart Ok, this has been changed in the new version. Now only one of the conditions can be satisfied. I think it would be good to make it even more explicit by using `else if` everywhere above (except in the first `if` of course) and `else` below, in which case you also won't need to initialise `base`.

File: drivers/soundwire/bus.c [hunk 6]
```diff
@@ -1082,12 +1082,125 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
 	return ret;
 }
 
+static int sdw_slave_set_frequency(struct sdw_slave *slave)
+{
+	u32 mclk_freq = slave->bus->prop.mclk_freq;
+	u32 curr_freq = slave->bus->params.curr_dr_freq >> 1;
+	u8 base = 0;
+	u8 scale_index;
+	u32 scale;
```
Comment (lyakh) [inline drivers/soundwire/bus.c]: I'd say `scale` should be just an `unsigned int`

File: drivers/soundwire/bus.c (hunk 5 above)
Comment (kv2019i) [reply drivers/soundwire/bus.c:1131]: @lyakh @plbossart  Hmm I can understand this now, thanks. 

I think the if-else would be in order now that mclk_freq is set in each if branch, so the others cannot be possible be hit anymore. 

Furthermore, if the known BIOS Issue is only with 38.4Mhz, I wonder do we need the "% mclk_freq" division check at all, if this is indeed a simple mapping. I.e. could we just do:

```
switch (mclk_freq) {
 case 19200000:
        base = SDW_SCP_BASE_CLOCK_19200000_HZ;
        break;
 case 24000000:
       /* as above */
 case 38400000:
        mclk_freq = 19200000;
        base = SDW_SCP_BASE_CLOCK_19200000_HZ;
        break;
   default: 
	dev_err(&slave->dev,
			"MCLK %d is not multiple of bus curr_freq %d\n",
			mclk_freq, curr_freq);
		return -EINVAL;
 }
```

... this would be much more obvious to the reader (i.e. a simple mapping).

Comment (bardliao) [reply drivers/soundwire/bus.c:1131]: > @lyakh @plbossart Hmm I can understand this now, thanks.
> 
> I think the if-else would be in order now that mclk_freq is set in each if branch, so the others cannot be possible be hit anymore.
> 
> Furthermore, if the known BIOS Issue is only with 38.4Mhz, I wonder do we need the "% mclk_freq" division check at all, if this is indeed a simple mapping. I.e. could we just do:
> 
> ```
> switch (mclk_freq) {
>  case 19200000:
>         base = SDW_SCP_BASE_CLOCK_19200000_HZ;
>         break;
>  case 24000000:
>        /* as above */
>  case 38400000:
>         mclk_freq = 19200000;
>         base = SDW_SCP_BASE_CLOCK_19200000_HZ;
>         break;
>    default: 
> 	dev_err(&slave->dev,
> 			"MCLK %d is not multiple of bus curr_freq %d\n",
> 			mclk_freq, curr_freq);
> 		return -EINVAL;
>  }
> ```
> 
> ... this would be much more obvious to the reader (i.e. a simple mapping).

I am not sure if the 38400000 case is an issue or not. But IMO we should support all valid mclk_freq. A platform (don't need to be Intel platform) may use say 12000000 mclk_freq which is not in the list but can be informed with base = 24000000 with a proper scale. So I think using % here is necessary.

**Comment by plbossart** [comment]
@lyakh @kv2019i @bardliao updated with if-else if cascade. A switch case ends-up being rather long with 3 values per root frequency, it's simpler to use a % to take care of the variations.

File: drivers/soundwire/bus.c (hunk 6 above)
Comment (plbossart) [reply drivers/soundwire/bus.c]: yes, fixed

Comment (plbossart) [reply drivers/soundwire/bus.c]: both fixed
